### PR TITLE
Feat: change forgotpassword activity architecture to mvp using kotlin

### PR DESCRIPTION
**Pull request by amitiwary999**
Fixed issue #801

### CODE REVIEW

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordPresenter.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+class ForgotPasswordPresenter : ForgotPasswordView {
+
+    var authResponseCall: Call<ForgotPasswordResponse>? = null
+
+    override fun cancelSignup() {
+        authResponseCall?.cancel()
+    }
+
+    override fun signup(email: String, url: String, isPersonalServerChecked: Boolean, context: Context, listener: ForgotPasswordView.onFinishListener) {
+        if (email.isEmpty()) {
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordPresenter.kt]: This all code should go in Interactor. You haven't even made an interactor. All the business logic should be in interactor/model and not in presenter.

**Comment by chiragw15** [comment]
Follow #793 and implement it similarly.

**Comment by amitiwary999** [comment]
@chiragw15 @mayank408  review.

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -43,7 +43,7 @@
             </intent-filter>
         </activity>
         <activity
-            android:name=".activities.ForgotPasswordActivity"
+            android:name=".forgotpassword.ForgotPasswordActivity"
```
Comment (chiragw15) [inline app/src/main/AndroidManifest.xml]: P in uppercase

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import android.view.MenuItem
+import android.view.View
+import kotlinx.android.synthetic.main.activity_forgot_password.*
+import org.fossasia.susi.ai.R
+import org.fossasia.susi.ai.helper.AlertboxHelper
+import org.fossasia.susi.ai.helper.Constant
+import org.fossasia.susi.ai.login.LoginActivity
+
+
+
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt]: remove empty lines

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * Created by meeera on 6/7/17.
+ */
+class ForgotPasswordActivity : AppCompatActivity (), ForgotPasswordView {
+
+    var forgetPasswordPresenter: IForgotPasswordPresenter?= null
+    var progressDialog: ProgressDialog? = null
+
+    override fun invalidUrl() {
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt]: move methods starting from `invalidUrl` to `wrongEmail` after `onOptionsItemSelected`

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_forgot_password)
+        title = getString(R.string.forgot_pass_activity)
+        showUrl()
+        hideUrl()
+        cancelSignUp()
+        signUp()
+
+        if(savedInstanceState != null) {
+            forgot_email.editText!!.setText(savedInstanceState.getCharSequenceArray(Constant.SAVED_STATES)!![0].toString())
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt]: remove !!  replace with ?

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            if (savedInstanceState.getBoolean(Constant.SERVER)) {
+               input_url.visibility = View.VISIBLE
+            } else {
+                input_url.visibility = View.GONE
+            }
+        }
+
+        if (supportActionBar != null) {
+            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt]: here also

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    }
+
+    fun cancelSignUp() {
+        progressDialog?.setOnCancelListener {
+            forgetPasswordPresenter?.cancelSignup()
+            reset_button.isEnabled = true
+        }
+    }
+    fun signUp() {
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt]: add line before this. Also signUp is a little confusing. Name something else

File: app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        val values = arrayOf<CharSequence>(forgot_email.getEditText()?.getText().toString())
+        outState.putCharSequenceArray(Constant.SAVED_STATES, values)
+        outState.putBoolean(Constant.SERVER, personal_server.isChecked())
+    }
+
+
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotpassword/ForgotPasswordActivity.kt]: remove extra empty lines

**Comment by amitiwary999** [comment]
@chiragw15  done

**Comment by mayank408** [comment]
@amitiwary999 Resolves the conflicts pls

**Comment by chiragw15** [comment]
Also change `ForgotPasswordView` to `IForgotPasswordView`. Rest everything is fine. Good work.

File: app/src/main/java/org/fossasia/susi/ai/forgotPassword/ForgotPasswordInteractor.kt [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                } else {
+                    listener.failure("${response.code()}" + context.getString(R.string.error), response.message(), context.getString(R.string.ok), Color.BLUE)
+                }
+                listener.hideProgress()
+            }
+
+            override fun onFailure(call: Call<ForgotPasswordResponse>, t: Throwable) {
+                t.printStackTrace()
+                if (t is UnknownHostException) {
+                    listener.failure(context.getString(R.string.unknown_host_exception), "${t.message}", context.getString(R.string.retry), Color.RED)
```
Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/forgotPassword/ForgotPasswordInteractor.kt]: Using t.message gives error

File: app/src/main/java/org/fossasia/susi/ai/forgotPassword/IForgotPasswordPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+/**
+ * Created by meeera on 7/7/17.
+ */
+interface IForgotPasswordPresenter {
+
+    fun onAttach(forgotPasswordView: ForgotPasswordView)
+    fun signup(email: String, url: String, isPersonalServerChecked: Boolean, context: Context)
+    fun cancelSignup()
+}
```
Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/forgotPassword/IForgotPasswordPresenter.kt]: Why? You used it in  presenter but not call it from activity.

File: app/src/main/java/org/fossasia/susi/ai/forgotPassword/ForgotPasswordInteractor.kt (hunk 9 above)
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotPassword/ForgotPasswordInteractor.kt]: use .toString()

File: app/src/main/java/org/fossasia/susi/ai/forgotPassword/IForgotPasswordPresenter.kt (hunk 10 above)
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotPassword/IForgotPasswordPresenter.kt]: Ohh nice. Thank you. Please change it there also.

Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/forgotPassword/IForgotPasswordPresenter.kt]: Check for signup activity as well.

**Comment by mariobehling** [comment]
Please ensure Circle-CI is building.

**Comment by amitiwary999** [comment]
@mariobehling done.

**Pull request merged**